crate/server/src/routes/google_cse/operations.rs

```diff
@@ -0,0 +1,206 @@
+use std::sync::Arc;
+
+use actix_web::HttpRequest;
+use base64::{engine::general_purpose, Engine};
+use clap::crate_version;
+use cosmian_kmip::kmip::{
+    kmip_data_structures::{KeyWrappingData, KeyWrappingSpecification},
+    kmip_objects::ObjectType,
+    kmip_types::{self, CryptographicAlgorithm, EncodingOption},
+};
+use cosmian_kms_utils::crypto::symmetric::create_symmetric_key;
+use serde::{Deserialize, Serialize};
+
+use super::GoogleCseConfig;
+use crate::{
+    core::operations::{unwrap_key, wrap_key},
+    result::KResult,
+    routes::google_cse::jwt::validate_tokens,
+    KMSServer,
+};
+
+#[derive(Serialize, Debug)]
+pub struct StatusResponse {
+    pub server_type: String,
+    pub vendor_id: String,
+    pub version: String,
+    pub name: String,
+    pub operations_supported: Vec<String>,
+}
+
+pub fn get_status() -> StatusResponse {
+    StatusResponse {
+        server_type: "KACLS".to_string(),
+        vendor_id: "Cosmian".to_string(),
+        version: crate_version!().to_string(),
+        name: "Cosmian KMS".to_string(),
+        operations_supported: vec![
+            "wrap".to_string(),
+            "unwrap".to_string(),
+            "digest".to_string(),
+            // "privilegedunwrap".to_string(),
+            // "privatekeydecrypt".to_string(),
+            // "privatekeysign".to_string(),
+            // "privilegedprivatekeydecrypt".to_string(),
+        ],
+    }
+}
+
+#[derive(Deserialize, Debug)]
+pub struct WrapRequest {
+    pub authentication: String,
+    pub authorization: String,
+    pub key: String,
+    pub reason: String,
+}
+
+#[derive(Serialize, Debug)]
+pub struct WrapResponse {
+    pub wrapped_key: String,
+}
+
+/// Returns encrypted Data Encryption Key (DEK) and associated data.
+///
+/// See [doc](https://developers.google.com/workspace/cse/reference/wrap) and
+/// for more details, see [Encrypt & decrypt data](https://developers.google.com/workspace/cse/guides/encrypt-and-decrypt-data)
+pub async fn wrap(
+    req_http: HttpRequest,
+    wrap_request: WrapRequest,
+    cse_config: &Option<GoogleCseConfig>,
+    kms: &Arc<KMSServer>,
+) -> KResult<WrapResponse> {
+    let database_params = kms.get_sqlite_enc_secrets(&req_http)?;
+
+    let application = if wrap_request.reason.contains("Meet") {
+        "meet"
+    } else {
+        "drive"
+    };
+
+    // let roles: &[&str] = if application == "meet" {
+    //     &["writer", "upgrader"]
+    // } else {
+    //     &["writer", "upgrader"]
+    // };
```

Comment:
```suggestion

```

Reply:
Yes. I will need to review this before we merge. Google documentation is not super clear on the subject

Reply:
Done